src/helper/remote_config/asm_data_listener.cpp

```diff
@@ -0,0 +1,99 @@
+// Unless explicitly stated otherwise all files in this repository are
+// dual-licensed under the Apache-2.0 License or BSD-3-Clause License.
+//
+// This product includes software developed at Datadog
+// (https://www.datadoghq.com/). Copyright 2021 Datadog, Inc.
+#include "asm_data_listener.hpp"
+#include "../json_helper.hpp"
+#include "exception.hpp"
+#include <optional>
+#include <rapidjson/document.h>
+
+void dds::remote_config::asm_data_listener::on_update(const config &config)
+{
+    std::unordered_map<std::string, dds::service_config::rule_data> rules_data =
+        {};
+    rapidjson::Document serialized_doc;
+    if (!json_helper::get_json_base64_encoded_content(
+            config.contents, serialized_doc)) {
+        throw error_applying_config("Invalid config contents");
+    }
+
+    auto rules_data_itr = json_helper::get_field_of_type(
+        serialized_doc, "rules_data", rapidjson::kArrayType);
+    if (!rules_data_itr) {
+        throw error_applying_config("Invalid config json contents: "
+                                    "rules_data key missing or invalid");
+    }
+
+    for (rapidjson::Value::ConstValueIterator itr =
+             rules_data_itr.value()->value.Begin();
+         itr != rules_data_itr.value()->value.End(); ++itr) {
+        if (!itr->IsObject()) {
+            throw error_applying_config("Invalid config json contents: "
+                                        "rules_data entry invalid");
+        }
+
+        auto id_itr =
+            json_helper::get_field_of_type(itr, "id", rapidjson::kStringType);
+        auto type_itr =
+            json_helper::get_field_of_type(itr, "type", rapidjson::kStringType);
+        auto data_itr =
+            json_helper::get_field_of_type(itr, "data", rapidjson::kArrayType);
+        if (!id_itr || !type_itr || !data_itr) {
+            throw error_applying_config(
+                "Invalid config json contents: "
+                "rules_data missing a field or field is invalid");
+        }
+
+        auto rule = rules_data.find(id_itr.value()->value.GetString());
+        if (rule == rules_data.end()) { // New rule
+            dds::service_config::rule_data new_rule_data = {
+                id_itr.value()->value.GetString(),
+                type_itr.value()->value.GetString()};
+            extract_data(data_itr.value(), new_rule_data);
+            if (!new_rule_data.data.empty()) {
+                rules_data.insert(
+                    {id_itr.value()->value.GetString(), new_rule_data});
+            }
+        } else {
+            extract_data(data_itr.value(), rule->second);
+        }
+    }
+
+    service_config_->set_rules_data(std::move(rules_data));
```

Comment:
Unfortunately this won't do, the configuration merge needs to be passed to the engine using `update_rule_data`, so you'll need to figure out if you want to merge a `rapidjson::document` into another `rapidjson::document` and then use the existing convenience functions to convert to `parameter`, or use intermediate structures and write a new conversion function to `parameter`.